src/include/main/db_main.h

```diff
@@ -0,0 +1,102 @@
+#pragma once
+
+#include <memory>
+#include <unordered_map>
+#include <utility>
+#include "catalog/catalog.h"
+#include "common/action_context.h"
+#include "common/stat_registry.h"
+#include "network/terrier_server.h"
+#include "settings/settings_param.h"
+#include "transaction/transaction_manager.h"
+
+namespace terrier {
+
+namespace settings {
+class SettingsManager;
+class SettingsTests;
+}  // namespace settings
+
+/**
+ * The DBMain Class holds all the singleton pointers. It has the full knowledge
+ * of the whole database systems and serves as a global context of the system.
+ * *Only the settings manager should be able to access the DBMain object.*
+ */
+class DBMain {
+ public:
+  DBMain() = default;
+
+  /**
+   * The constructor of DBMain
+   * @param param_map a map stores setting values
+   */
+  explicit DBMain(std::unordered_map<settings::Param, settings::ParamInfo> &&param_map)
+      : param_map_(std::move(param_map)) {}
+
+  /**
+   * This function boots the backend components.
+   * It initializes the following components in the following order:
+   *    SettingsManager
+   *    Garbage Collector
+   *    Catalog
+   *    Worker Pool
+   *    Logging
+   *    Stats
+   */
+  void Init();
+
+  /**
+   * Boots the traffic cop and networking layer, starts the server loop.
+   * It will block until server shuts down.
+   */
+  void Run();
+
+  /**
+   * Shuts down the server.
+   * It is worth noting that in normal cases, terrier will shut down and return from Run().
+   * So, use this function only when you want to shutdown the server from code.
+   * For example, in the end of unit tests when you want to shut down your test server.
+   */
+  void ForceShutdown();
+
+  // TODO(Weichen): Use unique ptr is enough.
+  /**
+   * Basic empty callbacks used by settings manager
+   * @param old_value the old value of corresponding setting
+   * @param new_value the new value of corresponding setting
+   * @param action_context action context for empty callback
+   */
+  void KnobCallback(void *old_value, void *new_value, const std::shared_ptr<common::ActionContext> &action_context);
+
+  /**
+   * Buffer pool size callback used by settings manager
+   * @param old_value old value of buffer pool size
+   * @param new_value new value of buffer pool size
+   * @param action_context action context for changing buffer pool size
+   */
+  void BufferPoolSizeCallback(void *old_value, void *new_value,
```

Comment:
DBMain seems like a really weird place to stick callbacks for Settings Manager actions.

Reply:
But it is the only place where all the components are accessible...

Reply:
Is there a hierarchy that makes sense without us having have to define action callbacks in DBMain? I'm envisioning potentially dozens or hundreds some day and having them all in DBMain seems incorrect.